ChangeLog.d/fix-crl-revocationDate.txt

```diff
@@ -0,0 +1,6 @@
+Bugfix
+   * A CRL's "revocationDate" entry field is on longer checked to be in the
+     past. This brings the implementation in line with RFC 5280. Note that
+     this also is a security fix in environments where the local clock cannot
+     be trusted (e.g., in an Intel SGX enclave). Reported by Raoul Strackx &
```

Comment:
That's an interesting point. As you've noted when adding tests (thanks for that!), currently we have a `config.h` option `MBEDTLS_HAVE_TIME_DATE` whose documentation reads:

```
 * System has time.h, time(), and an implementation for
 * mbedtls_platform_gmtime_r() (see below).
 * The time needs to be correct (not necessarily very accurate, but at least
 * the date should be correct). This is used to verify the validity period of
 * X.509 certificates.
 *
 * Comment if your system does not have a correct clock.
```
Perhaps this documentation should be made more explicit:
```
 * \warning You should comment this out if the system's clock could be controlled by an attacker.
```

I'm mentioning this because my initial reaction was "but on systems where the local clock can't be trusted, people should just undefine `MBEDTLS_HAVE_TIME_DATE`" - then I checked if the documentation was actually making that clear, and perhaps not enough.

Reply:
“people should just undefine `MBEDTLS_HAVE_TIME_DATE`” That depends on your threat model. Undefining `MBEDTLS_HAVE_TIME_DATE` means that no certificate will ever be considered to have expired.

Reply:
I agree description could be more clear. The same applies for `MBEDTLS_HAVE_TIME`: when the system's clock can't be trusted, it also cannot be trusted for measuring time differences.

Reply:
Good point @gilles-peskine-arm but if you leave `MBEDTLS_HAVE_TIME_DATE` defined, things like SSL caches may never expire.

Reply:
Ok, after thinking a bit more about this, I think I was wrong to suggest placing the ChangeLog entry under "Bugfix". As the PR description says, when `MBEDTLS_HAVE_TIME_DATE` is unset, we would simply never consider any certificate as revoked. I'm not sure how I overlooked this before, but that's enough to qualify as a security issue.

I'm now thinking the ChangeLog entry should look more like:

```
Security
   * When checking X.509 CRLs, a certificate was only considered as revoked if its revocationDate was in the past according to the local clock if available. In particular, on builds without MBEDTLS_HAVE_TIME_DATE, certificates were never considered as revoked. On builds with MBEDTLS_HAVE_TIME_DATE, an attacker able to control the local clock (for example, an untrusted OS attacking a secure enclave) could prevent revocation of certificates via CRLs. Fixed by no longer checking the revocationDate field, in accordance with RFC 5280.  Reported and fixed by Raoul Strackx & Jethro Beekman.
```